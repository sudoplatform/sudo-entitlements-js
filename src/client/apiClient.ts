import {
  ApiClientManager,
  DefaultApiClientManager,
} from '@sudoplatform/sudo-api-client'
import * as SudoCommon from '@sudoplatform/sudo-common'
import { NormalizedCacheObject } from 'apollo-cache-inmemory'
import AWSAppSyncClient from 'aws-appsync'
import {
  EntitlementsSet,
  GetEntitlementsDocument,
  GetEntitlementsQuery,
  RedeemEntitlementsDocument,
  RedeemEntitlementsMutation,
} from '../gen/graphqlTypes'
/**
 * AppSync wrapper to use to invoke Entitlements Service APIs.
 */
export class ApiClient {
  private readonly client: AWSAppSyncClient<NormalizedCacheObject>

  public constructor(apiClientManager?: ApiClientManager) {
    const clientManager =
      apiClientManager ?? DefaultApiClientManager.getInstance()

    this.client = clientManager.getClient({ disableOffline: true })
  }

  public async getEntitlements(): Promise<EntitlementsSet | null> {
    const result = await this.client.query<GetEntitlementsQuery>({
      query: GetEntitlementsDocument,
      variables: {},
      fetchPolicy: 'no-cache',
    })

    return result.data.getEntitlements ?? null
  }

  public async redeemEntitlements(): Promise<EntitlementsSet> {
    const result = await this.client.mutate<RedeemEntitlementsMutation>({
      mutation: RedeemEntitlementsDocument,
      variables: {},
      fetchPolicy: 'no-cache',
    })

    if (result.data) {
      return result.data.redeemEntitlements ?? null
    } else {
      throw new SudoCommon.FatalError(
        'redeemEntitlements did not return any result.',
      )
    }
  }
}
