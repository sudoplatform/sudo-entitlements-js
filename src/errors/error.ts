/**
 * The identity token contained multiple recognized claims identifying
 * different entitlements sets on redemption.
 */
export class AmbiguousEntitlementsError extends Error {
  constructor() {
    super('Ambiguous entitlements.')
    this.name = 'AmbiguousEntitlementsError'
  }
}

/**
 * The identity token contained no claims identifying the external
 * ID of the user
 */
export class NoExternalIdError extends Error {
  constructor() {
    super('No external ID identifiable from token.')
    this.name = 'NoExternalIdError'
  }
}

/**
 * The identity token contained no claims identifying the billing group
 * of the user
 */
export class NoBillingGroupError extends Error {
  constructor() {
    super('No billing group identifiable from token.')
    this.name = 'NoBillingGroupError'
  }
}

/**
 * The entitlements set referenced in the identity token was
 * not found.
 */
export class EntitlementsSetNotFoundError extends Error {
  constructor() {
    super('Entitlements set referenced in token was not found.')
    this.name = 'EntitlementsSetNotFoundError'
  }
}

/**
 * The entitlements sequence referenced in the identity token was
 * not found.
 */
export class EntitlementsSequenceNotFoundError extends Error {
  constructor() {
    super('Entitlements sequence referenced in token was not found.')
    this.name = 'EntitlementsSequenceNotFoundError'
  }
}
