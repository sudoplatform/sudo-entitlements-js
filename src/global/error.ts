/**
 * Indicates the operation requires the user to be signed in but the user is
 * currently not signed in.
 */
export class NotSignedInError extends Error {
  constructor() {
    super('Not signed in.')
    this.name = 'NotSignedInError'
  }
}

/**
 * An unexpected error was encountered. This may result from programmatic error
 * and is unlikely to be user recoverable.
 */
export class FatalError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'FatalError'
  }
}
