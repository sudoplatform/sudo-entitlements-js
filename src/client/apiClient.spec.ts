import { ApiClientManager } from '@sudoplatform/sudo-api-client'
import {
  InvalidTokenError,
  NoEntitlementsError,
  ServiceError,
} from '@sudoplatform/sudo-common'
import { NormalizedCacheObject } from 'apollo-cache-inmemory'
import { NetworkStatus } from 'apollo-client'
import AWSAppSyncClient from 'aws-appsync'
import { anything, instance, mock, reset, verify, when } from 'ts-mockito'
import { AmbiguousEntitlementsError } from '../errors/error'
import {
  EntitlementsConsumption,
  EntitlementsSet,
  GetEntitlementsConsumptionQuery,
  GetEntitlementsQuery,
  RedeemEntitlementsMutation,
} from '../gen/graphqlTypes'
import { ApiClient } from './apiClient'

describe('ApiClient test suite', () => {
  const mockApiClientManager = mock<ApiClientManager>()
  const mockAWSAppSyncClient = mock<AWSAppSyncClient<NormalizedCacheObject>>()

  let apiClient: ApiClient

  beforeEach(() => {
    reset(mockApiClientManager)
    reset(mockAWSAppSyncClient)

    when(mockApiClientManager.getClient(anything())).thenReturn(
      instance(mockAWSAppSyncClient),
    )

    apiClient = new ApiClient(instance(mockApiClientManager))
  })

  const now = Date.now()
  const entitlementsSet: EntitlementsSet = {
    createdAtEpochMs: now,
    updatedAtEpochMs: now,
    version: 1,
    name: 'jest',
    entitlements: [{ name: 'jest-1', value: 1 }],
  }

  const entitlementsConsumption: EntitlementsConsumption = {
    entitlements: {
      entitlementsSetName: entitlementsSet.name,
      version: 1.00001,
      entitlements: entitlementsSet.entitlements,
    },
    consumption: [
      {
        name: entitlementsSet.entitlements[0].name,
        value: entitlementsSet.entitlements[0].value,
        consumed: 1,
        available: entitlementsSet.entitlements[0].value - 1,
      },
    ],
  }

  describe('getEntitlements tests', () => {
    it('should return data if returned', async () => {
      when(
        mockAWSAppSyncClient.query<GetEntitlementsQuery>(anything()),
      ).thenResolve({
        data: { getEntitlements: entitlementsSet },
        loading: false,
        networkStatus: NetworkStatus.ready,
        stale: false,
      })

      await expect(apiClient.getEntitlements()).resolves.toEqual(
        entitlementsSet,
      )

      verify(mockAWSAppSyncClient.query(anything())).once()
    })

    it('should throw a FatalError if no data and no errors returned', async () => {
      when(mockAWSAppSyncClient.query(anything())).thenResolve({
        data: undefined,
        errors: undefined,
        loading: false,
        networkStatus: NetworkStatus.ready,
        stale: false,
      })

      await expect(
        apiClient.getEntitlements(),
      ).rejects.toThrowErrorMatchingSnapshot()

      verify(mockAWSAppSyncClient.query(anything())).once()
    })

    it.each`
      result
      ${undefined}
      ${null}
    `('should return null if $result returned', async ({ result }) => {
      when(
        mockAWSAppSyncClient.query<GetEntitlementsQuery>(anything()),
      ).thenResolve({
        data: { getEntitlements: result },
        loading: false,
        networkStatus: NetworkStatus.ready,
        stale: false,
      })

      await expect(apiClient.getEntitlements()).resolves.toEqual(null)

      verify(mockAWSAppSyncClient.query(anything())).once()
    })
    it.each`
      code                                  | error
      ${'sudoplatform.NoEntitlementsError'} | ${new NoEntitlementsError()}
      ${'sudoplatform.InvalidTokenError'}   | ${new InvalidTokenError()}
      ${'sudoplatform.ServiceError'}        | ${new ServiceError('graphql-error')}
    `('should map $code error', async ({ code, error }) => {
      when(mockAWSAppSyncClient.query(anything())).thenResolve({
        data: undefined,
        loading: false,
        networkStatus: NetworkStatus.ready,
        stale: false,
        errors: [
          {
            errorType: code,
            message: 'graphql-error',
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
          } as any,
        ],
      })

      await expect(apiClient.getEntitlements()).rejects.toThrow(error)
    })
  })

  describe('getEntitlementsConsumption tests', () => {
    it('should return data if returned', async () => {
      when(
        mockAWSAppSyncClient.query<GetEntitlementsConsumptionQuery>(anything()),
      ).thenResolve({
        data: { getEntitlementsConsumption: entitlementsConsumption },
        loading: false,
        networkStatus: NetworkStatus.ready,
        stale: false,
      })

      await expect(apiClient.getEntitlementsConsumption()).resolves.toEqual(
        entitlementsConsumption,
      )

      verify(mockAWSAppSyncClient.query(anything())).once()
    })

    it('should throw a FatalError if no data and no errors returned', async () => {
      when(mockAWSAppSyncClient.query(anything())).thenResolve({
        data: undefined,
        errors: undefined,
        loading: false,
        networkStatus: NetworkStatus.ready,
        stale: false,
      })

      await expect(
        apiClient.getEntitlementsConsumption(),
      ).rejects.toThrowErrorMatchingSnapshot()

      verify(mockAWSAppSyncClient.query(anything())).once()
    })
    it.each`
      code                                  | error
      ${'sudoplatform.NoEntitlementsError'} | ${new NoEntitlementsError()}
      ${'sudoplatform.InvalidTokenError'}   | ${new InvalidTokenError()}
      ${'sudoplatform.ServiceError'}        | ${new ServiceError('graphql-error')}
    `('should map $code error', async ({ code, error }) => {
      when(mockAWSAppSyncClient.query(anything())).thenResolve({
        data: undefined,
        loading: false,
        networkStatus: NetworkStatus.ready,
        stale: false,
        errors: [
          {
            errorType: code,
            message: 'graphql-error',
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
          } as any,
        ],
      })

      await expect(apiClient.getEntitlementsConsumption()).rejects.toThrow(
        error,
      )
    })
  })

  describe('redeemEntitlements tests', () => {
    it('should throw a FatalError if no data and no errors returned', async () => {
      when(
        mockAWSAppSyncClient.mutate<RedeemEntitlementsMutation>(anything()),
      ).thenResolve({})

      await expect(
        apiClient.redeemEntitlements(),
      ).rejects.toThrowErrorMatchingSnapshot()

      verify(mockAWSAppSyncClient.mutate(anything())).once()
    })

    it('should return data if returned', async () => {
      when(
        mockAWSAppSyncClient.mutate<RedeemEntitlementsMutation>(anything()),
      ).thenResolve({
        data: { redeemEntitlements: entitlementsSet },
      })

      await expect(apiClient.redeemEntitlements()).resolves.toEqual(
        entitlementsSet,
      )

      verify(mockAWSAppSyncClient.mutate(anything())).once()
    })

    it.each`
      result
      ${undefined}
      ${null}
    `('should return null if $result returned', async ({ result }) => {
      when(
        mockAWSAppSyncClient.mutate<RedeemEntitlementsMutation>(anything()),
      ).thenResolve({
        data: { redeemEntitlements: result },
      })

      await expect(apiClient.redeemEntitlements()).resolves.toEqual(null)

      verify(mockAWSAppSyncClient.mutate(anything())).once()
    })

    it.each`
      code                                                      | error
      ${'sudoplatform.entitlements.AmbiguousEntitlementsError'} | ${new AmbiguousEntitlementsError()}
      ${'sudoplatform.InvalidTokenError'}                       | ${new InvalidTokenError()}
      ${'sudoplatform.ServiceError'}                            | ${new ServiceError('graphql-error')}
    `('should map $code error', async ({ code, error }) => {
      when(
        mockAWSAppSyncClient.mutate<RedeemEntitlementsMutation>(anything()),
      ).thenResolve({
        errors: [
          {
            errorType: code,
            message: 'graphql-error',
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
          } as any,
        ],
      })
      await expect(apiClient.redeemEntitlements()).rejects.toThrow(error)
    })
  })
})
