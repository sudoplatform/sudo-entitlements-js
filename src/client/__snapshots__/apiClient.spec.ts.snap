// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiClient test suite redeemEntitlements tests should throw a FatalError if no data returned 1`] = `"redeemEntitlements did not return any result."`;
